### PR TITLE
En_Raf (Carnivorous Lily Pad) OK and documented, object_raf documented

### PR DESCRIPTION
**Pull request by tom-overton**
Before opening this PR, ensure the following:
- `./format.sh` was run to apply standard formatting.
- `make` successfully builds a matching ROM.
- No new compiler warnings were introduced during the build process.
    - Can be verified locally by running `tools/warnings_count/check_new_warnings.sh`
- New variables & functions should follow standard naming conventions.
- Comments and variables have correct spelling.
---

I don't think the names in this actor are really the best, but they're about as good as I can get by myself; please feel free to comment on the bad ones and suggest improvements. I also didn't document the damage table because I believe it's irrelevant? If you feed it an explosive, it will blow up, and otherwise the damage table isn't even looked at.

### CODE REVIEW

**Review by hensldm — APPROVED** [review]
Seems good to me. Can't think of any improvements at the moment.

File: src/overlays/actors/ovl_En_Raf/z_en_raf.c
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Raf/z_en_raf.c:185]: This is one pixel at a time since it's an rgba16, so I think to singular instead.

File: src/overlays/actors/ovl_En_Raf/z_en_raf.c
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Raf/z_en_raf.c:204]: Does the pad have to go exactly here?

File: src/overlays/actors/ovl_En_Raf/z_en_raf.c
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Raf/z_en_raf.c:272]: Might as well go in the function, I suppose.

File: src/overlays/actors/ovl_En_Raf/z_en_raf.c
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Raf/z_en_raf.c:337]: Well that's weird.

File: src/overlays/actors/ovl_En_Raf/z_en_raf.c
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Raf/z_en_raf.c:568]:
```suggestion
 * Plays the convulsing animation and sets the lily pad's switch flag to prevent it from
```

File: src/overlays/actors/ovl_En_Raf/z_en_raf.c
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Raf/z_en_raf.c:611]: I suppose this is related to the texture dimensions?

File: src/overlays/actors/ovl_En_Raf/z_en_raf.c
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Raf/z_en_raf.c:663]: I suppose there is a question of whether it can be dead if it gets better... how about "dormant"?

File: src/overlays/actors/ovl_En_Raf/z_en_raf.c
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Raf/z_en_raf.c:831]: I know this was discussed in Discord, I think it essentially stretches them a bit in a particular direction, right? it undoes all the limb rotations, scales/stretches, then puts them all back to get it in the same place.

File: src/overlays/actors/ovl_En_Raf/z_en_raf.c
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Raf/z_en_raf.c:883]: I think OoT has moved away from "particle" to just calling these "special effects" or something, would recommend investigating the current terminology to keep us in sync. (I don't have a strong opinion on which is better, so happy to go the OoT way.)

File: src/overlays/actors/ovl_En_Raf/z_en_raf.h
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Raf/z_en_raf.h:13]: Nice, keeping up the Hokuto no Ken references...

File: src/overlays/actors/ovl_En_Raf/z_en_raf.h
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Raf/z_en_raf.h:14]: I suppose this is never used, since afaik there's only a few of these in the entire game?

File: src/overlays/actors/ovl_En_Raf/z_en_raf.h
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Raf/z_en_raf.h:44]: I think just call this "bobOffset` or something... otherwise given how it's next to `heightDiffFromPlayer`, the inevitable question is "who's Bob?".

File: src/overlays/actors/ovl_En_Raf/z_en_raf.h
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Raf/z_en_raf.h:51]: Maybe `RotY` over `YRot`?

File: src/overlays/actors/ovl_En_Raf/z_en_raf.h
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Raf/z_en_raf.h:52]: Better make this `mainType`, I think, or it disappears into all the other types.

File: src/overlays/actors/ovl_En_Raf/z_en_raf.h
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Raf/z_en_raf.h:59]: Do these come from OoT? I remember several OoT actors (Gohma, Phantom Ganon) do similar things.

**Review by EllipticEllipsis — COMMENTED** [review]
Good job on this, I think most of the names are fine, but I have a few thoughts for where they might be improved.

File: src/overlays/actors/ovl_En_Raf/z_en_raf.h
Comment (tom-overton) [reply src/overlays/actors/ovl_En_Raf/z_en_raf.h:13]: I'll be 100% honest, it actually came from Bloodborne: https://www.youtube.com/watch?v=4F54phkQgiU

But that in itself might be a reference to something else. Either way, this is probably getting replaced if I switch from `DeadIdle` to `Dormant`, but it was fun while it lasted.

File: assets/xml/objects/object_raf.xml
Comment (tom-overton) [reply assets/xml/objects/object_raf.xml:44]: Whoops, no it is not. Thanks for catching it.

File: src/overlays/actors/ovl_En_Raf/z_en_raf.c
Comment (tom-overton) [reply src/overlays/actors/ovl_En_Raf/z_en_raf.c:204]: Yes, it has to be right there to get the stack to line up. I'm guessing the `GlobalContext *` temp was just there for some reason instead of at the top.

File: src/overlays/actors/ovl_En_Raf/z_en_raf.c
Comment (tom-overton) [reply src/overlays/actors/ovl_En_Raf/z_en_raf.c:611]: Yep, it's 16 * 32. Should I express it as such, like this? I would do this with all texture dimension stuff.
```c
if (this->petalClearPixelFirstPassIndex < (16 * 32))
```

File: src/overlays/actors/ovl_En_Raf/z_en_raf.h
Comment (tom-overton) [reply src/overlays/actors/ovl_En_Raf/z_en_raf.h:14]: Yeah, there are only 7 of these things, and every single one is of type `EN_RAF_TYPE_NORMAL`.

File: src/overlays/actors/ovl_En_Raf/z_en_raf.c
Comment (tom-overton) [reply src/overlays/actors/ovl_En_Raf/z_en_raf.c:337]: Yeah, I have no clue why they did this either. It's definitely some sort of sanity check, but it seems excessive.

File: src/overlays/actors/ovl_En_Raf/z_en_raf.c
Comment (tom-overton) [reply src/overlays/actors/ovl_En_Raf/z_en_raf.c:831]: Yeah, that's what it does. I'll make the comment a little more clear as to what exactly is happening here.

File: src/overlays/actors/ovl_En_Raf/z_en_raf.h
Comment (tom-overton) [reply src/overlays/actors/ovl_En_Raf/z_en_raf.h:59]: Gohma uses `decayingProgress`, whereas Phantom Ganon uses `work[GND_BODY_DECAY_INDEX]`. However, Gohma has `sClearPixelTableFirstPass` and `sClearPixelTableSecondPass`, which is where I got the names for these things.

File: src/overlays/actors/ovl_En_Raf/z_en_raf.c
Comment (tom-overton) [reply src/overlays/actors/ovl_En_Raf/z_en_raf.c:883]: Hmmm, well, most of the bosses call them "effects" because they use `globalCtx->specialEffects`. I can find one instance of an actor in OoT using "effect" *without* using `globalCtx->specialEffects`, which is En_G_Switch. Meanwhile, there are a few more instances of "particle" being used for this purpose instead, including for actors whose code for handling these things looks nearly-identical to Raf's. Unfortunately, there isn't a ton of consistency in OoT either. My gut says to leave this for now and we can decide on the particle vs effect nomenclature later, but what do you think?

File: src/overlays/actors/ovl_En_Raf/z_en_raf.c
Comment (EllipticEllipsis) [reply src/overlays/actors/ovl_En_Raf/z_en_raf.c:608]: Yes, once zeldaret/ZAPD#233 is in we'll have macros for them anyway though.

File: src/overlays/actors/ovl_En_Raf/z_en_raf.c
Comment (EllipticEllipsis) [reply src/overlays/actors/ovl_En_Raf/z_en_raf.c:204]: Yeah, just unusual, that's all. Presumably the inline data forces the stack to be exactly this.

File: src/overlays/actors/ovl_En_Raf/z_en_raf.c
Comment (tom-overton) [reply src/overlays/actors/ovl_En_Raf/z_en_raf.c:608]: Ah, well, let me just do them manually for now then. If I forget to come back to this once the ZAPD change is merged, feel free to remind me.

File: src/overlays/actors/ovl_En_Raf/z_en_raf.c
Comment (EllipticEllipsis) [reply src/overlays/actors/ovl_En_Raf/z_en_raf.c:334]: More than that, it just doesn't make any sense... I'd be tempted to `@bug` it except that it doesn't do anything.

File: src/overlays/actors/ovl_En_Raf/z_en_raf.c
Comment (EllipticEllipsis) [reply src/overlays/actors/ovl_En_Raf/z_en_raf.c:881]: Yeah, let's not worry about it for now then... I did think there'd been a push for consensus on effects. 🤔

File: src/overlays/actors/ovl_En_Raf/z_en_raf.c
Comment (tom-overton) [reply src/overlays/actors/ovl_En_Raf/z_en_raf.c:204]: Yeah, that's probably it. I remember having a devil of a time getting the stack in this function to work because of this pad.

File: src/overlays/actors/ovl_En_Raf/z_en_raf.c
Comment (tom-overton) [reply src/overlays/actors/ovl_En_Raf/z_en_raf.c:881]: I did a quick discord search for "particle" and I didn't find anything recently. So if there *was* a consensus, maybe it was a while ago?

File: src/overlays/actors/ovl_En_Raf/z_en_raf.h
Comment (EllipticEllipsis) [reply src/overlays/actors/ovl_En_Raf/z_en_raf.h]: https://knowyourmeme.com/memes/you-are-already-dead-omae-wa-mou-shindeiru

File: src/overlays/actors/ovl_En_Raf/z_en_raf.c
Comment (tom-overton) [reply src/overlays/actors/ovl_En_Raf/z_en_raf.c:335]: Do you want me to slap a comment above this saying it's a pointless check?

Comment (EllipticEllipsis) [reply src/overlays/actors/ovl_En_Raf/z_en_raf.c:335]: Yes, that's probably worth doing, if it was totally obvious it wouldn't still be there... I hope.

File: src/overlays/actors/ovl_En_Raf/z_en_raf.c
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Raf/z_en_raf.c:185]: I suppose the other thing to note here is that this function will *only* work with a texture that uses 16-bit pixels, hence why `texture` is a `u16*` here.

File: src/overlays/actors/ovl_En_Raf/z_en_raf.c
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Raf/z_en_raf.c:335]: I think I'd just say
```suggestion
            // This check is pointless, since EnRaf is never in the explosive category.
```

**Review by EllipticEllipsis — APPROVED** [review]

**Review by Kenix3 — APPROVED** [review]
We have generally been preferring to use the term "effects" over "particles," but no reason to hold up this PR in my opinion. We have a lot of other places in MM that are using the term "particle right now."

Other than that, the names aren't bad.

**Pull request merged**
